### ISSUE TITLE
AbstractService.SendHttpRequestMessagePagedAsync with filter callback

### ISSUE THREAD

**Issue by oscar54321** [dougdellolio/coinbasepro-csharp#252]
Hello Guys,

This is a proposed new feature.

This is a default request responsible for pagination. Sometimes it is not possible to tell how many pages is required. For example, I want to retrieve all the historical trades to reconstruct the latest bar. I can't tell in in advance how many pages in historical trades sequence are going to be there, so I would like to keep retrieving page after page until I reach a point that is older than the start of my last bar. Hence, a callback is most efficient method. Callback receives: page index and latest page content and returns true if the page of interest has been reached at what point the loop can exit. Legacy function can provide callback to exit on the number of pages. 

```
     protected async Task<IList<IList<T>>> SendHttpRequestMessagePagedAsync<T>(
            HttpMethod httpMethod,
            string uri,
            string content = null,
            int numberOfPages = 0)
```
```
   protected async Task<IList<IList<T>>> SendHttpRequestMessagePagedAsync<T>(
           HttpMethod httpMethod,
           string uri,
           Func<int, IList<T>, bool> isLastPageCriteria,
           string content = null)
        {
            Log.Debug("REST {HttpMethod} {Uri} {Content} ", httpMethod, uri, content);

            var pagedList = new List<IList<T>>();

            var httpResponseMessage = await SendHttpRequestMessageAsync(httpMethod, uri, content);
            var contentBody = await httpClient.ReadAsStringAsync(httpResponseMessage).ConfigureAwait(false);

            var firstPage = JsonConfig.DeserializeObject<IList<T>>(contentBody);

            pagedList.Add(firstPage);

            if (!httpResponseMessage.Headers.TryGetValues("cb-after", out var firstPageAfterCursorId))
            {
                return pagedList;
            }

            if (isLastPageCriteria(0, firstPage))
                return pagedList;

            var subsequentPages = await GetAllSubsequentPages<T>(uri, firstPageAfterCursorId.First(), isLastPageCriteria);

            pagedList.AddRange(subsequentPages);

            return pagedList;
        }


 private async Task<IList<IList<T>>> GetAllSubsequentPages<T>(
           string uri,
           string firstPageAfterCursorId,
           Func<int, IList<T>, bool> isLastPageCriteria)
        {
            var pagedList = new List<IList<T>>();
            var subsequentPageAfterHeaderId = firstPageAfterCursorId;

            // pageNumber = 0 was the first page
            int pageNumber = 1;

            int requestPerSecond = 0;

            while (true)
            {
                if (requestPerSecond >= 5)
                {
                    await Task.Delay(1000);
                    requestPerSecond = 0;
                }

                Log.Debug("REST {HttpMethod} {Uri} {PageAfter} ", HttpMethod.Get, uri, subsequentPageAfterHeaderId);

                async Task<HttpResponseMessage> call()
                {
                    return await SendHttpRequestMessageAsync(HttpMethod.Get, uri + $"&after={subsequentPageAfterHeaderId}").ConfigureAwait(false);
                }

                HttpResponseMessage subsequentHttpResponseMessage = await ExecuteRetry(call);

                if (!subsequentHttpResponseMessage.Headers.TryGetValues("cb-after", out var cursorHeaders))
                {
                    break;
                }

                subsequentPageAfterHeaderId = cursorHeaders.First();

                var subsequentContentBody = await httpClient.ReadAsStringAsync(subsequentHttpResponseMessage).ConfigureAwait(false);
                var page = JsonConfig.DeserializeObject<IList<T>>(subsequentContentBody);

                pagedList.Add(page);

                if (isLastPageCriteria(pageNumber, page))
                    break;

                pageNumber++;

                requestPerSecond++;
            }

            return pagedList;
        }
```

**Comment by dougdellolio** [comment]
hi @oscar54321. there is a default value of 0 assigned to the `numberOfPages` parameter on paginated requests. I would suggest not passing in any value here as this will create a request for all available pages.  please let me know if this doesn't solve the problem.

**Comment by oscar54321** [comment]
Hi Doug,

Let me clarify, I am trying to implement the suggested pattern for bootstrapping the realitme feed and I want my feed to build candles. So, I subscribe for trade updates using web socket and begin caching web socket match messages. Then, I begin retrieving historical trades (this is a paginated request without parameters) in order to build current candle. Once I received trades up to the candle start time, I first apply historical trades and then replay websocket cached portion discarding trades that have been retrieved from history. After this process is complete, last candle is in sync with web socket and can be continuously updated from the websocket stream. 

In this scenario, I must to be able to retrieve trades up to current bar start time (I know start time as the .Now rounded by time frame resolution, for example, if current time is 10:14 am and time frame is 5 min, the bar that I would want to build starts at 00:10 am and ends 00:15 am. So, for this scenario I want to retrieve trades up to to certain point in the past. It is obvious, that if current time is very close to candle start (for example, time is 00:10:03 am (3 seconds worth) , there probably will be very few trades that might be returned within a single historical trades response. However, if it is say 00:14:15, I might need to go a bunch of pages to collect all trades belonging to the current bar. It is impossible to say how many in general case - it depends on market activity.

If there is a gap in trade ID in in websocket, I need to retrieve trades that are missing from sequence and need to be replayed. In this case, I should go back and retrieve trades up to certain trade ID. I theoretically could estimate the number of trades I need to go back, knowing the current last trade ID and the trade ID I have, however, it is again not possible to say how many pages back from now I need to go through because the market can generate a bunch of newer trades any moment.

Bottom line, it is not possible to know without going through each page when you need to stop.

**Comment by dougdellolio** [comment]
I want to make sure I am understanding what you are trying to do. You are building realtime candle and using the websocket feed to populate the data (as is the recommended solution), however, the docs state that the matches channel may drop messages (trades) so you must revert to the REST API to find the missing trades.


> Please note that messages can be dropped from this channel. By using the heartbeat channel you can track the last trade id and fetch trades that you missed from the REST API.


Are you currently subscribing to the heartbeat channel for this? You will know from the heartbeat channel immediately that a trade was missed based on LastTradeId. I guess where I am still not clear is if you can calculate how many trades have passed, why would you need to know how many pages are required or what page the missing trade is on? You can request how many items you want per page. So for example, you know 50 trades have passed since the missing trade, you can request for 51 items (using the limit param) for 1 page as GetTradesAsync() returns in descending time order. Or if 200 trades have passed, you can request for 100 items for 3 pages (100 item limit per page). You would be able to search through the returned results to find your missing trade and fill in the gap. apologies if my understanding is incorrect but hopefully this helps, I haven't implemented anything like this.

**Comment by oscar54321** [comment]
Doug,

To build realtime candles, I combine data from both historical trades REST and and websocket to build realtime candle, because candles can start in the past relative to the moment when I try to start building it. For example, its noon now and daily candle started 12 hours ago. Ask yourself, how many pages of historical trades is that? 50? 500? 50000? In order to retrieve history, I may need to go quite a bit of pages (that have a limit of 300 trades per page) back in history. And, most importantly, it is not possible to say how many pages I need to retrieve beforehand because of a combination of two reasons a) I don't know how many trades happened between this moment and candle start and/or b) market keeps generating trades, so even if I new both, the last trade ID and the ID of first trade that happened  during candle, by the time my request is handled by exchange, there could be more trades and more pages generated. So, only by looking at the content of last retrieved page can I decide that it is time to stop. That's why callback for paginated requests makes a lot of sense to me.

*On a side note, heatbeats don't really send last trade ID despite stating so in the docs: last trade IDs are always zero there. I checked this in browser and in web socket stream.

**Comment by dougdellolio** [comment]
so while I think this is a valid problem, I am not convinced this should be added right now. This feels like a solution to a specific problem that you might be having. I would say if this is crucial to what you are trying to do you should fork the project and update it as you wish. I will keep this issue open and the request in my planner for the future should any other users come forward requesting something like this be done.

**Comment by oscar54321** [comment]
Doug, it's embarassing but I am not really good at github and forking. Would it be ok to ask for some very minor tweaks to framework to make it more extensible? I could have derived from this client it and override some services substituting my own extensions. 

a) CoinbaseProClient: store local instances such as clock, query builder etc. in protected properties so that derived class could access them.
b) CoinbaseProClient: ideally make service properties "protected set"
c) AbstractService: expose as protected httpRequestMessageService, httpClient.

```
 public CoinbaseProClient(
            IAuthenticator authenticator,
            IHttpClient httpClient,
            bool sandBox = false)
        {
            var clock = new Clock();
            var httpRequestMessageService = new HttpRequestMessageService(authenticator, clock, sandBox);
            var createWebSocketFeed = (Func<IWebSocketFeed>)(() => new WebSocketFeed(sandBox));
            var queryBuilder = new QueryBuilder();

            AccountsService = new AccountsService(httpClient, httpRequestMessageService);
            CoinbaseAccountsService = new CoinbaseAccountsService(httpClient, httpRequestMessageService);
            OrdersService = new OrdersService(httpClient, httpRequestMessageService, queryBuilder);
            PaymentsService = new PaymentsService(httpClient, httpRequestMessageService);
            WithdrawalsService = new WithdrawalsService(httpClient, httpRequestMessageService, queryBuilder);
            DepositsService = new DepositsService(httpClient, httpRequestMessageService);
            ProductsService = new ProductsService(httpClient, httpRequestMessageService, queryBuilder);
            CurrenciesService = new CurrenciesService(httpClient, httpRequestMessageService);
            FillsService = new FillsService(httpClient, httpRequestMessageService);
            FundingsService = new FundingsService(httpClient, httpRequestMessageService, queryBuilder);
            ReportsService = new ReportsService(httpClient, httpRequestMessageService);
            UserAccountService = new UserAccountService(httpClient, httpRequestMessageService);
            StablecoinConversionsService = new StablecoinConversionsService(httpClient, httpRequestMessageService);
            FeesService = new FeesService(httpClient, httpRequestMessageService);
            ProfilesService = new ProfilesService(httpClient, httpRequestMessageService);
            WebSocket = new WebSocket.WebSocket(createWebSocketFeed, authenticator, clock);

            Log.Information("CoinbaseProClient constructed");
        }
```

**Comment by dougdellolio** [comment]
No worries @oscar54321. Forking this project might be the best option for you as it sounds like there are some important changes needed for your specific project. 

To fork you can press the "Fork" button at the top right of the repo's main page. This will create a copy of the code on your account. GitHub Docs has a really good "Getting Started" with forking if you need. https://docs.github.com/en/free-pro-team@latest/github/getting-started-with-github/fork-a-repo.  After forking, you can clone your version of the repo locally and change the code as you please. You will also be able to reference your version of the code in the project you are working on, hopefully closing all of the gaps you have run into.

Please let me know if I can help any further
